README.md
# Grd2Shp_Xagg

[![PyPI](https://img.shields.io/pypi/v/grd2shp_xagg.svg)](https://pypi.org/project/grd2shp_xagg/)
[![Status](https://img.shields.io/pypi/status/grd2shp_xagg.svg)](https://pypi.org/project/grd2shp_xagg/)
[![Python
Version](https://img.shields.io/pypi/pyversions/grd2shp_xagg)](https://pypi.org/project/grd2shp_xagg)
[![License](https://img.shields.io/pypi/l/grd2shp_xagg)](https://opensource.org/licenses/MIT)

[![Read the documentation at
https://grd2shp_xagg.readthedocs.io/](https://img.shields.io/readthedocs/grd2shp_xagg/latest.svg?label=Read%20the%20Docs)](https://grd2shp_xagg.readthedocs.io/)
[![Tests](https://github.com/rmcd-mscb/grd2shp_xagg/workflows/Tests/badge.svg)](https://github.com/rmcd-mscb/grd2shp_xagg/actions?workflow=Tests)
[![Codecov](https://codecov.io/gh/rmcd-mscb/grd2shp_xagg/branch/main/graph/badge.svg)](https://codecov.io/gh/rmcd-mscb/grd2shp_xagg)

[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

## Features2

- TODO

## Requirements

- TODO

## Installation

You can install _Grd2Shp_Xagg_ via [pip](https://pip.pypa.io/) from
[PyPI](https://pypi.org/):

```console
$ pip install grd2shp_xagg
```

## Usage

Please see the [Command-line Reference](#usage) for details.

## Contributing

Contributions are very welcome. To learn more, see the [Contributor
Guide](CONTRIBUTING.rst).

## License

Distributed under the terms of the [MIT
license](https://opensource.org/licenses/MIT), _Grd2Shp_Xagg_ is free
and open source software.

## Issues

If you encounter any problems, please [file an
issue](https://github.com/rmcd-mscb/grd2shp_xagg/issues) along with a
detailed description.

### Develop

```console
$ conda env create -f environment.yml
$ conda develop -n {{cookiecutter.project_name}} src
$ conda activate {{cookiecutter.project_name}}
$ pip install -r requirements.dev
```

It is important to get [pre-commit]() enabled on the project, to ensure
that certain standards are always met on a git commit. With several of
these, it might fail if files are changed, but it will change them, and
trying the commit a second time will actually work.

### Git hook configuration

```console
$ pre-commit install --install-hooks
```

### Testing

[Nox]() is used for testing everything, with several sessions built-in.
To run the full suite of tests, simply use:

```console
$ nox
```

The different sessions are:

    - `pre-commit` -- validates that the [pre-commit]() checks all come back clean.
    - `safety` -- validates the [Safety]() of all production dependencies.
    - `mypy` -- validates the type-hints for the application using [mypy]().
    - `tests` -- runs all [pytest]() tests.
    - `typeguard` -- runs all [pytest]() tests, validates with [Typeguard]().
    - `xdoctest` -- runs any and all documentation examples with [xdoctest]().
    - `docs-build` -- builds the full set of generated API docs with [Sphinx]().

These can be run individually with the following command:

```console
$ nox -s <session>
```

Replace `<session>` with the name of the session give above, i.e.:

```console
$ nox -s mypy
```

You can also simply run [pytest]() tests, by using the command:

```console
$ pytest tests
```

### Dependencies

Production dependencies are duplicated, in both `requirements.txt` and
`environment.yml` due to how [conda]() does not work with the
`requirements.txt` file. It is necessary for both files to be updated as
dependencies are added.

Development dependencies are contained in `requirements.dev`.

### Version Management

    - Warning:
      Do not set the initial version to `0.0.0`. This will break, due to
      an unknown bug in [Bump2version]() at this time.

The projects made by this cookiecutter use [Bump2version]() for version
management. The default version that the project starts with is a
developmental version, `0.0.1-dev0`. In github, this should be
auto-incremented on each commit to the next dev build number. To manage
the version changes yourself, you can use the [Bump2version]() command:

```console
$ bump2version <part>
```

Where `<part>` is one of:

    - `major`
    - `minor`
    - `patch`
    - `build`

  <!-- end list -->

    - Note:
      This makes a `dev` version, which does not write a tag into git. It
      is just useful for development purposes and not the version that is
      recommended for a release version. The version string will be
      formatted as: `<major>.<minor>.<patch>-dev<build>`

To do a production release, use the command:

```console
$ bump2version --tag release
```

This will add a tag in the git repository noting the version.

    - Note:
      The version string for this will be: `<major>.<minor>.<patch>`

The template supports Python 3.8 and 3.9.

## Credits

This project was generated from
[@hillc-usgs](https://github.com/hillc-usgs)'s [Pygeoapi Plugin
Cookiecutter](https://code.usgs.gov/wma/nhgf/pygeoapi-plugin-cookiecutter)
template.


HISTORY.rst
=======
History
=======

0.0.1-dev0 (2021-07-30)
------------------

* Some description here.


noxfile.py
"""Nox sessions."""
import shutil
import sys
from pathlib import Path
from textwrap import dedent

import nox
from nox.sessions import Session


package = "grd2shp_xagg"
python_versions = ["3.9", "3.8"]
nox.needs_version = ">= 2021.6.6"
nox.options.sessions = (
    "pre-commit",
    "safety",
    "mypy",
    "tests",
    "typeguard",
    "xdoctest",
    "docs-build",
)


def activate_virtualenv_in_precommit_hooks(session: Session) -> None:
    """Activate virtualenv in hooks installed by pre-commit.

    This function patches git hooks installed by pre-commit to activate the
    session's virtual environment. This allows pre-commit to locate hooks in
    that environment when invoked from git.

    Args:
        session: The Session object.
    """
    if session.bin is None:
        return

    virtualenv = session.env.get("VIRTUAL_ENV")
    if virtualenv is None:
        return

    hookdir = Path(".git") / "hooks"
    if not hookdir.is_dir():
        return

    for hook in hookdir.iterdir():
        if hook.name.endswith(".sample") or not hook.is_file():
            continue

        text = hook.read_text()
        bindir = repr(session.bin)[1:-1]  # strip quotes
        if not (
            Path("A") == Path("a") and bindir.lower() in text.lower() or bindir in text
        ):
            continue

        lines = text.splitlines()
        if not (lines[0].startswith("#!") and "python" in lines[0].lower()):
            continue

        header = dedent(
            f"""\
            import os
            os.environ["VIRTUAL_ENV"] = {virtualenv!r}
            os.environ["PATH"] = os.pathsep.join((
                {session.bin!r},
                os.environ.get("PATH", ""),
            ))
            """
        )

        lines.insert(1, header)
        hook.write_text("\n".join(lines))


@nox.session(name="pre-commit", python="3.9")
def precommit(session: Session) -> None:
    """Lint using pre-commit."""
    args = session.posargs or ["run", "--all-files", "--show-diff-on-failure"]
    session.install(
        "black",
        "darglint",
        "flake8",
        "flake8-bandit",
        "flake8-bugbear",
        "flake8-docstrings",
        "flake8-rst-docstrings",
        "pep8-naming",
        "pre-commit",
        "pre-commit-hooks",
        "reorder-python-imports",
    )
    session.run("pre-commit", *args)
    if args and args[0] == "install":
        activate_virtualenv_in_precommit_hooks(session)


@nox.session(python="3.9")
def safety(session: Session) -> None:
    """Scan dependencies for insecure packages."""
    session.install("safety", ".")
    session.run("safety", "check", "--full-report")


@nox.session(python=python_versions)
def mypy(session: Session) -> None:
    """Type-check using mypy."""
    args = session.posargs or ["src", "tests", "docs/conf.py"]
    session.install(".")
    session.install("mypy", "pytest")
    session.run("mypy", *args)
    if not session.posargs:
        session.run("mypy", f"--python-executable={sys.executable}", "noxfile.py")


@nox.session(python=python_versions)
def tests(session: Session) -> None:
    """Run the test suite."""
    session.install(".")
    session.install("coverage[toml]", "pytest", "pygments")
    try:
        session.run("coverage", "run", "--parallel", "-m", "pytest", *session.posargs)
    finally:
        if session.interactive:
            session.notify("coverage", posargs=[])


@nox.session
def coverage(session: Session) -> None:
    """Produce the coverage report."""
    args = session.posargs or ["report"]

    session.install("coverage[toml]")

    if not session.posargs and any(Path().glob(".coverage.*")):
        session.run("coverage", "combine")

    session.run("coverage", *args)


@nox.session(python=python_versions)
def typeguard(session: Session) -> None:
    """Runtime type checking using Typeguard."""
    session.install(".")
    session.install("pytest", "typeguard", "pygments")
    session.run("pytest", f"--typeguard-packages={package}", *session.posargs)


@nox.session(python=python_versions)
def xdoctest(session: Session) -> None:
    """Run examples with xdoctest."""
    args = session.posargs or ["all"]
    session.install(".")
    session.install("xdoctest[colors]")
    session.run("python", "-m", "xdoctest", package, *args)


@nox.session(name="docs-build", python="3.9")
def docs_build(session: Session) -> None:
    """Build the documentation."""
    args = session.posargs or ["docs", "docs/_build"]
    session.install(".")
    session.install("sphinx", "sphinx-click", "sphinx-rtd-theme")

    build_dir = Path("docs", "_build")
    if build_dir.exists():
        shutil.rmtree(build_dir)

    session.run("sphinx-build", *args)


@nox.session(python="3.9")
def docs(session: Session) -> None:
    """Build and serve the documentation with live reloading on file changes."""
    args = session.posargs or ["--open-browser", "docs", "docs/_build"]
    session.install(".")
    session.install("sphinx", "sphinx-autobuild", "sphinx-click", "sphinx-rtd-theme")

    build_dir = Path("docs", "_build")
    if build_dir.exists():
        shutil.rmtree(build_dir)

    session.run("sphinx-autobuild", *args)
